atxi pagination param can cause slice OOB panic

```
panic: runtime error: slice bounds out of range

goroutine 8523 [running]:
github.com/ethereumproject/go-ethereum/core.GetAddrTxs.func1(0xc420164240, 0x8, 0x8, 0xc4710d7db0, 0x42, 0xc420164240)
	/Users/ia/gocode/src/github.com/ethereumproject/go-ethereum/core/atxi.go:374 +0x218
github.com/ethereumproject/go-ethereum/core.GetAddrTxs(0x4eb54c0, 0xc425687890, 0x4755566335d4bcf6, 0x4fc72826c4943649, 0xc442baaa33, 0x43d845, 0x46d182, 0x4cdd009, 0x1, 0x4cdd009, ...)
	/Users/ia/gocode/src/github.com/ethereumproject/go-ethereum/core/atxi.go:377 +0x54a
github.com/ethereumproject/go-ethereum/eth.(*PublicGethAPI).GetAddressTransactions(0xc4200bcab8, 0x4755566335d4bcf6, 0x4fc72826c4943649, 0x42baaa33, 0x43d845, 0x46d182, 0xc4367d9fea, 0x2, 0xc4367d9fec, 0x2, ...)
	/Users/ia/gocode/src/github.com/ethereumproject/go-ethereum/eth/api.go:1734 +0x374
reflect.Value.call(0xc420164000, 0xc4200bdaf8, 0x13, 0x4cddc5f, 0x4, 0xc4b71b0b60, 0x9, 0x9, 0x756ea19ac3b7f1, 0xc4b71b0b60, ...)
	/usr/local/Cellar/go/1.10.2/libexec/src/reflect/value.go:447 +0x969
reflect.Value.Call(0xc420164000, 0xc4200bdaf8, 0x13, 0xc4b71b0b60, 0x9, 0x9, 0x8, 0x8, 0x30)
	/usr/local/Cellar/go/1.10.2/libexec/src/reflect/value.go:308 +0xa4
github.com/ethereumproject/go-ethereum/rpc.(*Server).handle(0xc42afc93e0, 0x4eb17e0, 0xc4a6e7a420, 0x4eb7560, 0xc4fda1f3b0, 0xc42f2e0a20, 0xc4779fa200, 0x18, 0xc425ad0738)
	/Users/ia/gocode/src/github.com/ethereumproject/go-ethereum/rpc/server.go:330 +0x68c
github.com/ethereumproject/go-ethereum/rpc.(*Server).exec(0xc42afc93e0, 0x4eb17e0, 0xc4a6e7a420, 0x4eb7560, 0xc4fda1f3b0, 0xc42f2e0a20)
	/Users/ia/gocode/src/github.com/ethereumproject/go-ethereum/rpc/server.go:352 +0x1b0
github.com/ethereumproject/go-ethereum/rpc.(*Server).serveRequest.func2(0xc4666f7540, 0xc42afc93e0, 0xc4f0cd1650, 0x4eb7560, 0xc4fda1f3b0, 0xc422293c60, 0x1, 0x1, 0x0)
	/Users/ia/gocode/src/github.com/ethereumproject/go-ethereum/rpc/server.go:223 +0xad
created by github.com/ethereumproject/go-ethereum/rpc.(*Server).serveRequest
	/Users/ia/gocode/src/github.com/ethereumproject/go-ethereum/rpc/server.go:218 +0x291
```

--- a/cmd/geth/build_atxi_cmd.go
+++ b/cmd/geth/build_atxi_cmd.go
@@ -1,11 +1,12 @@
 package main
 
 import (
+	"math"
+
 	"github.com/ethereumproject/go-ethereum/core"
 	"github.com/ethereumproject/go-ethereum/ethdb"
 	"github.com/ethereumproject/go-ethereum/logger/glog"
 	"gopkg.in/urfave/cli.v1"
-	"math"
 )
 
 var buildAddrTxIndexCommand = cli.Command{

--- a/core/atxi.go
+++ b/core/atxi.go
@@ -304,12 +304,12 @@ func GetAddrTxs(db ethdb.Database, address common.Address, blockStartN uint64, b
 		err = errWithReason(errAtxiInvalidUse, "Address transactions list signature requires 'kind of' param to be empty string or [s|c] prefix (eg. both, standard, or contract)")
 		return
 	}
-	if paginationStart < 0 {
-		paginationStart = 0
-	}
-	if paginationStart > paginationEnd {
+	if paginationStart > 0 && paginationEnd > 0 && paginationStart > paginationEnd {
 		err = errWithReason(errAtxiInvalidUse, "Pagination start must be less than or equal to pagination end params")
 		return
+	}
+	if paginationStart < 0 {
+		paginationStart = 0
 	}
 
 	// Have to cast to LevelDB to use iterator. Yuck.

--- a/core/atxi.go
+++ b/core/atxi.go
@@ -367,8 +367,12 @@ func GetAddrTxs(db ethdb.Database, address common.Address, blockStartN uint64, b
 		}
 		if paginationStart < 0 {
 			paginationStart = 0
+		} else if paginationStart > len(s) {
+			paginationStart = len(s)
 		}
 		if paginationEnd < 0 {
+			paginationEnd = len(s)
+		} else if paginationEnd > len(s) {
 			paginationEnd = len(s)
 		}
 		return s[paginationStart:paginationEnd]

--- a/core/atxi.go
+++ b/core/atxi.go
@@ -5,11 +5,6 @@ import (
 	"encoding/binary"
 	"errors"
 	"fmt"
-	"github.com/ethereumproject/go-ethereum/common"
-	"github.com/ethereumproject/go-ethereum/core/types"
-	"github.com/ethereumproject/go-ethereum/ethdb"
-	"github.com/ethereumproject/go-ethereum/logger"
-	"github.com/ethereumproject/go-ethereum/logger/glog"
 	"math"
 	"os"
 	"os/signal"
@@ -17,10 +12,17 @@ import (
 	"strings"
 	"syscall"
 	"time"
+
+	"github.com/ethereumproject/go-ethereum/common"
+	"github.com/ethereumproject/go-ethereum/core/types"
+	"github.com/ethereumproject/go-ethereum/ethdb"
+	"github.com/ethereumproject/go-ethereum/logger"
+	"github.com/ethereumproject/go-ethereum/logger/glog"
 )
 
 var (
 	errAtxiNotEnabled = errors.New("atxi not intialized")
+	errAtxiInvalidUse = errors.New("invalid parameters passed to ATXI")
 
 	txAddressIndexPrefix = []byte("atx-")
 	txAddressBookmarkKey = []byte("ATXIBookmark")
@@ -288,18 +290,33 @@ func (bc *BlockChain) GetATXIBuildProgress() (*AtxiProgressT, error) {
 
 // GetAddrTxs gets the indexed transactions for a given account address.
 // 'reverse' means "oldest first"
-func GetAddrTxs(db ethdb.Database, address common.Address, blockStartN uint64, blockEndN uint64, direction string, kindof string, paginationStart int, paginationEnd int, reverse bool) []string {
+func GetAddrTxs(db ethdb.Database, address common.Address, blockStartN uint64, blockEndN uint64, direction string, kindof string, paginationStart int, paginationEnd int, reverse bool) (txs []string, err error) {
+	errWithReason := func(e error, s string) error {
+		return fmt.Errorf("%v: %s", e, s)
+	}
+
+	// validate params
 	if len(direction) > 0 && !strings.Contains("btf", direction[:1]) {
-		glog.Fatal("Address transactions list signature requires direction param to be empty string or [b|t|f] prefix (eg. both, to, or from)")
+		err = errWithReason(errAtxiInvalidUse, "Address transactions list signature requires direction param to be empty string or [b|t|f] prefix (eg. both, to, or from)")
+		return
 	}
 	if len(kindof) > 0 && !strings.Contains("bsc", kindof[:1]) {
-		glog.Fatal("Address transactions list signature requires 'kind of' param to be empty string or [s|c] prefix (eg. both, standard, or contract)")
+		err = errWithReason(errAtxiInvalidUse, "Address transactions list signature requires 'kind of' param to be empty string or [s|c] prefix (eg. both, standard, or contract)")
+		return
+	}
+	if paginationStart < 0 {
+		paginationStart = 0
+	}
+	if paginationStart > paginationEnd {
+		err = errWithReason(errAtxiInvalidUse, "Pagination start must be less than or equal to pagination end params")
+		return
 	}
 
 	// Have to cast to LevelDB to use iterator. Yuck.
 	ldb, ok := db.(*ethdb.LDBDatabase)
 	if !ok {
-		return nil
+		err = errWithReason(errors.New("internal interface error; please file a bug report"), "could not cast eth db to level db")
+		return nil, nil
 	}
 
 	// This will be the returnable.
@@ -351,8 +368,9 @@ func GetAddrTxs(db ethdb.Database, address common.Address, blockStartN uint64, b
 		atxis = append(atxis, atxi{blockN: bn, tx: tx})
 	}
 	it.Release()
-	if e := it.Error(); e != nil {
-		glog.Fatalln(e)
+	err = it.Error()
+	if err != nil {
+		return
 	}
 
 	handleSorting := func(s sortableAtxis) sortableAtxis {
@@ -365,20 +383,16 @@ func GetAddrTxs(db ethdb.Database, address common.Address, blockStartN uint64, b
 				s[i], s[j] = s[j], s[i]
 			}
 		}
-		if paginationStart < 0 {
-			paginationStart = 0
-		} else if paginationStart > len(s) {
+		if paginationStart > len(s) {
 			paginationStart = len(s)
 		}
-		if paginationEnd < 0 {
-			paginationEnd = len(s)
-		} else if paginationEnd > len(s) {
+		if paginationEnd < 0 || paginationEnd > len(s) {
 			paginationEnd = len(s)
 		}
 		return s[paginationStart:paginationEnd]
 	}
-
-	return handleSorting(atxis).TxStrings()
+	txs = handleSorting(atxis).TxStrings()
+	return
 }
 
 // RmAddrTx removes all atxi indexes for a given tx in case of a transaction removal, eg.

--- a/core/blockchain_test.go
+++ b/core/blockchain_test.go
@@ -982,23 +982,30 @@ func TestFastVsFullChainsATXI(t *testing.T) {
 			}
 		}
 
-		out := GetAddrTxs(db, addr1, 0, 0, "", "", -1, -1, false)
+		out, _ := GetAddrTxs(db, addr1, 0, 0, "", "", -1, -1, false)
 		if len(out) != 3 {
 			t.Errorf("[%d] got: %v, want: %v", i, len(out), 3)
 		}
-		out = GetAddrTxs(db, addr1, 0, 0, "from", "", -1, -1, false)
+
+		// method should return an error if pagination params are invalid
+		_, err = GetAddrTxs(db, addr1, 0, 0, "", "", 2, 1, false)
+		if err == nil {
+			t.Errorf("[%d] got: %v, want: %v", i, err, errAtxiInvalidUse)
+		}
+
+		out, _ = GetAddrTxs(db, addr1, 0, 0, "from", "", -1, -1, false)
 		if len(out) != 2 {
 			t.Errorf("[%d] got: %v, want: %v", i, len(out), 2)
 		}
-		out = GetAddrTxs(db, addr1, 0, 0, "to", "", -1, -1, false)
+		out, _ = GetAddrTxs(db, addr1, 0, 0, "to", "", -1, -1, false)
 		if len(out) != 1 {
 			t.Errorf("[%d] got: %v, want: %v", i, len(out), 1)
 		}
-		out = GetAddrTxs(db, addr2, 0, 0, "", "", -1, -1, false)
+		out, _ = GetAddrTxs(db, addr2, 0, 0, "", "", -1, -1, false)
 		if len(out) != 3 {
 			t.Errorf("[%d] got: %v, want: %v", i, len(out), 3)
 		}
-		out = GetAddrTxs(db, addr2, 3, 3, "", "", -1, -1, false)
+		out, _ = GetAddrTxs(db, addr2, 3, 3, "", "", -1, -1, false)
 		if len(out) != 1 {
 			t.Errorf("[%d] got: %v, want: %v", i, len(out), 1)
 		}
@@ -1076,14 +1083,14 @@ func TestRmAddrTx(t *testing.T) {
 		t.Fatalf("failed to process block %d: %v", res.Index, res.Error)
 	}
 
-	out := GetAddrTxs(db, addr1, 0, 0, "", "", -1, -1, false)
+	out, _ := GetAddrTxs(db, addr1, 0, 0, "", "", -1, -1, false)
 	if len(out) != 3 {
 		t.Errorf("got: %v, want: %v", len(out), 3)
 	}
 	if err := RmAddrTx(db, t1); err != nil {
 		t.Fatal(err)
 	}
-	out = GetAddrTxs(db, addr1, 0, 0, "", "", -1, -1, false)
+	out, _ = GetAddrTxs(db, addr1, 0, 0, "", "", -1, -1, false)
 	if len(out) != 2 {
 		t.Errorf("got: %v, want: %v", len(out), 2)
 	}
@@ -1363,9 +1370,9 @@ func testChainTxReorgs(t *testing.T, db ethdb.Database, withATXI bool) {
 	if !withATXI {
 		return
 	}
-	txsh1 := GetAddrTxs(db, addr1, 0, 0, "", "", -1, -1, false)
-	txsh2 := GetAddrTxs(db, addr2, 0, 0, "", "", -1, -1, false)
-	txsh3 := GetAddrTxs(db, addr3, 0, 0, "", "", -1, -1, false)
+	txsh1, _ := GetAddrTxs(db, addr1, 0, 0, "", "", -1, -1, false)
+	txsh2, _ := GetAddrTxs(db, addr2, 0, 0, "", "", -1, -1, false)
+	txsh3, _ := GetAddrTxs(db, addr3, 0, 0, "", "", -1, -1, false)
 
 	allAtxis := txsh1
 	allAtxis = append(allAtxis, txsh2...)

--- a/core/database_util_test.go
+++ b/core/database_util_test.go
@@ -395,13 +395,13 @@ func TestAddrTxStorage(t *testing.T) {
 		t.Errorf("want: %v, got: %v", 7, count)
 	}
 
-	out := GetAddrTxs(db, from2, 0, 0, "", "", -1, -1, false)
+	out, _ := GetAddrTxs(db, from2, 0, 0, "", "", -1, -1, false)
 	if len(out) != 3 {
 		t.Errorf("want: %v, got: %v", 3, len(out))
 	}
 
 	// Test pagination and reverse
-	outReverse := GetAddrTxs(db, from2, 0, 0, "", "", -1, -1, true)
+	outReverse, _ := GetAddrTxs(db, from2, 0, 0, "", "", -1, -1, true)
 	if len(outReverse) != 3 {
 		t.Errorf("want: %v, got: %v", 3, len(outReverse))
 	}
@@ -410,21 +410,21 @@ func TestAddrTxStorage(t *testing.T) {
 		t.Errorf("got: %v, want: %v", outReverse, out)
 	}
 	// pagination
-	outPag := GetAddrTxs(db, from2, 0, 0, "", "", 1, -1, false)
+	outPag, _ := GetAddrTxs(db, from2, 0, 0, "", "", 1, -1, false)
 	if len(outPag) != 2 {
 		t.Errorf("got: %v, want: %v", len(outPag), 2)
 	}
 
-	out = GetAddrTxs(db, from2, 0, 0, "", "c", -1, -1, false)
+	out, _ = GetAddrTxs(db, from2, 0, 0, "", "c", -1, -1, false)
 	if len(out) != 1 {
 		t.Errorf("got: %v, want: %v", len(out), 1)
 	}
-	out = GetAddrTxs(db, common.Address{}, 0, 0, "", "", -1, -1, false)
+	out, _ = GetAddrTxs(db, common.Address{}, 0, 0, "", "", -1, -1, false)
 	if len(out) != 1 {
 		t.Errorf("got: %v, want: %v", len(out), 1)
 	}
 
-	out = GetAddrTxs(db, from1, 314, 314, "", "", -1, -1, false)
+	out, _ = GetAddrTxs(db, from1, 314, 314, "", "", -1, -1, false)
 	if len(out) != 1 {
 		t.Errorf("want: %v, got: %v", 1, len(out))
 	} else {
@@ -447,7 +447,7 @@ func TestAddrTxStorage(t *testing.T) {
 		}
 	}
 
-	out = GetAddrTxs(db, from2to, 314, 314, "to", "", -1, -1, false)
+	out, _ = GetAddrTxs(db, from2to, 314, 314, "to", "", -1, -1, false)
 	if len(out) != 1 {
 		t.Errorf("want: %v, got: %v", 1, len(out))
 	} else {
@@ -474,7 +474,7 @@ func TestAddrTxStorage(t *testing.T) {
 			t.Errorf("got: %v, want: %v", f, from2)
 		}
 	}
-	out = GetAddrTxs(db, from2to, 314, 314, "from", "", -1, -1, false)
+	out, _ = GetAddrTxs(db, from2to, 314, 314, "from", "", -1, -1, false)
 	if len(out) != 0 {
 		t.Errorf("want: %v, got: %v", 0, len(out))
 	}

--- a/eth/api.go
+++ b/eth/api.go
@@ -1731,7 +1731,10 @@ func (api *PublicGethAPI) GetAddressTransactions(address common.Address, blockSt
 		blockEndN = 0
 	}
 
-	list = core.GetAddrTxs(atxi.Db, address, blockStartN, uint64(blockEndN.Int64()), toOrFrom, txKindOf, pagStart, pagEnd, reverse)
+	list, err = core.GetAddrTxs(atxi.Db, address, blockStartN, uint64(blockEndN.Int64()), toOrFrom, txKindOf, pagStart, pagEnd, reverse)
+	if err != nil {
+		return
+	}
 
 	// Since list is a slice, it can be nil, which returns 'null'.
 	// Should return empty 'array' if no txs found.

